src/utils/Task.ts

```diff
@@ -0,0 +1,92 @@
+export type PerfStats = {
+    realTime: number;
+    cpuTime: number;
+};
+
+type IncrementalFunc = (() => void);
+
+/**
+ * A class that can break down work into chunks that run within a given time budget per
+ * iteration. It allows the UI to update. It can also be cancelled.
+ */
+export class Task<T, S = {}> {
+    private result: T | undefined;
+    private cancelled: boolean = false;
+    private onComplete: ((res: T, stats: PerfStats) => void) | undefined;
+    private timeBudget: number = Infinity;
+    private startTime: number = new Date().getTime();
+    private cpuTime: number = 0;
+    private lastContext: S | null;
+
+    /**
+     * @param {number} timeBudget How much time in seconds can be spent per frame before the rest
+     * of the work is deferred to the next frame. Note that it will only stop once it has gone OVER
+     * this time budget (hopefully just by a little bit) so it is not a hard limit.
+     */
+    constructor(
+        iterator: IterableIterator<T | undefined>,
+        timeBudget: number,
+        withContext: ((instance: S | null, callback: IncrementalFunc) => void) = () => {},
+        maybeContext: (() => S | null) = () => {
+            return null;
+        }
+    ) {
+        this.timeBudget = timeBudget;
+
+        // This will continue running the `iterator` coroutine to continue the optimization process until
+        // either it finishes or we spent too long in the current frame, in which case, we schedule this
+        // function again in another `requestAnimationFrame` callback to continue it after the UI gets a
+        // chance to update.
+        const incrementalWork = () => {
+            const incrementalStartTime = new Date().getTime();
+
+            const existsTimeRemaining = () =>
+                (new Date().getTime() - incrementalStartTime) / 1000 < this.timeBudget;
+
+            withContext(this.lastContext, () => {
```

Comment:
I think the `withContext` thing is pretty specific to generator definitions. Maybe it'll be better to keep it out of `Task`, and then for our `generateSOSMC` use of `Task`, decorating the iterator, so we have something like:

```typescript
class GenerationIterator implements IterableIterator<Model | undefined> {
  private lastContext: GeneratorInstance;
  private iterator: IterableIterator<Model | undefined>;
  public next() {
    withContext(this.lastContext, () => {
        value = this.iterator.next();
    });
    this.lastContext = maybeContext();
    return value;
  }
}
```

Reply:
Ya, wasn't sure on how general this context thing is. Will update with your suggestion when I get home tonight! :)